### DISCUSSION TITLE
Dark mode in NextJS 13 without using next-themes?

### DISCUSSION THREAD

**Discussion by ajiteshchhatani** [vercel/next.js#50786] [Help]
### Summary

I was trying out the app dir in NextJS13 and setup a fresh project where I created a `navbar` component which had a `theme_button` button component to toggle between light and dark theme. I also created a `theme-provider` file to create a client side context.

`app/theme-provider.tsx`

```
"use client";

import { createContext, useState } from "react";

interface ThemeContextType {
  theme: string;
  changeTheme: (value: string) => void;
}

const defaultThemeValue: ThemeContextType = {
  theme: "light",
  changeTheme: (value: string) => {},
};

export const ThemeContext = createContext<ThemeContextType>(defaultThemeValue);

export default function ThemeProvider({
  children,
}: {
  children: React.ReactNode;
}) {
  const [theme, setTheme] = useState<string>(localStorage.getItem("theme") as string);

  const handleChangeTheme = (value: string) => {
    setTheme(value => value);
    localStorage.setItem("theme", value);
  };

  return (
    <ThemeContext.Provider value={{theme: theme, changeTheme: (theme: string) => handleChangeTheme(theme)}}>
      {children}
    </ThemeContext.Provider>
  );
}
```

`theme-button.tsx`
```
"use client";

import { useContext, useEffect, useState } from "react";
import { ThemeContext } from "./theme-provider";

export default function ThemeButton() {

  const [theme, setTheme] = useState<string>("light");

  const themeContextValue = useContext(ThemeContext);

  const handleThemeChange = () => {
    setTheme(theme => theme === "light" ? "dark" : "light");
    themeContextValue.changeTheme(theme)
  }

  return (
    <button
      className="flex items-center gap-2"
      onClick={handleThemeChange}
    >
      <img
        src="/moon-svgrepo-com.svg"
        alt="change theme icon"
        width={30}
        height={30}
      />Dark Mode
    </button>
  );
}
```

`app/layout.tsx`
```
import "./globals.css";
import { Inter } from "next/font/google";
import ThemeProvider from "./theme-provider";

const inter = Inter({ subsets: ["latin"] });

export const metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <ThemeProvider>
        {/**HOW TO READ Context here without making this a client component? Is it advisable to make this a client component? */}
        <body className={inter.className}>{children}</body>
      </ThemeProvider>
    </html>
  );
}
```

My question is what is the correct approach in reading context for values such as themes in the layout.tsx file so the entire body can be given the theme using the dark class approach of tailwindCSS?

Should we make the entire layout as client component? Is that even advisable?

Any help is appreciated!

**Comment by Hamza-31** [comment]
I'm also working on this dark mode feature with the latest NextJS 13 and TailwindCSS. In my case, I didn't have to use Context from React to create a ThemeProvider. So I only created two custom hooks: useLocalStorage and useColorMode. This first one allows to store the data (colorMode). The second hook allows to add and remove dark class to the body.

useLocalStorage.tsx
```
'use client'
import { useEffect, useState } from 'react';

const useLocalStorage = (key: string, initialValue:string) => {
const [value, setValue] = useState(() => {
if (typeof window !== 'undefined'){

	try {
		 
			const storedValue = window.localStorage.getItem(key);
			return storedValue ? JSON.parse(storedValue) : initialValue;
		
	} catch (error) {
	  console.error(`Error retrieving value from localStorage: ${error}`);
	}
}
return initialValue;
});

useEffect(() => {
if (typeof window !== 'undefined') {
try {
const serializedValue = JSON.stringify(value);
window.localStorage.setItem(key, serializedValue);
} catch (error) {
console.error(Error storing value in localStorage: ${error});
}
}
}, [key, value]);

return [value, setValue];
};

export default useLocalStorage;
```

useColorMode.tsx
```
'use client'
import React, { useEffect } from 'react'
import useLocalStorage from './useLocalStorage'

const useColorMode = () => {
const [colorMode, setColorMode] = useLocalStorage("color-mode", "light")

useEffect(() => {
	const className = "dark"
	const bodyClasses = window.document.body.classList
	colorMode === "dark" ? bodyClasses.add(className) : bodyClasses.remove(className)
}, [colorMode])
return [colorMode, setColorMode]
}

export default useColorMode
```

ThemeButton.tsx
```
'use client'

import useColorMode from "@/lib/useColorMode"
import { SunIcon, MoonIcon } from "@heroicons/react/24/solid"

const ThemeButton = () => {
const [colorMode, setColorMode] = useColorMode()
return (
<button
aria-label="Toggle Dark Mode"
type="button"
className="flex items-center justify-center rounded-lg p-2"
onClick={() => { setColorMode(colorMode === "dark" ? "light" : "dark") }}
>
{
colorMode === "dark" ? (
			) : (<MoonIcon className="h-5 w-5 text-dark-purple" />)
		}
	</button>
)
}

export default ThemeButton
```

In this approach I store the theme value in the client local storage and count on the clients to render their own theme. Since, in NextJS, we render React pages on the server, this causes flickering dark mode and also rendering the wrong dark/light mode icon. In real terms, the server rendered page is displayed on the client before the theme can be updated.

So I wonder if the Context Provider can :

- persist dark mode.
- prevent flickering behavior.
- handle version mismatch between client and server.

**Reply by ajiteshchhatani** [reply]
Hey, while this is an alternative solution where you're using custom hooks to save and manage the theme, are you actually applying the dark theme css on all client components instead of just applying it like so?

```
...
<body className={dark:background-black dark:color-white} />
..
```

My question even with your solution still is the same. How would you read the theme from localStorage in layout.jsx and apply it to the body entirely instead of applying the classes to all client components?

And yes I agree we will have flickering as body.layout.tsx would be rendered on the server then localStorage would be read to apply the theme. I think that problem will be there even with Context.

**Comment by Shorifpatwary** [comment]
# Try this code. 
### just import this component. no need to use  useContext hook or next theme. 
_____________________________________________________
`
"use client";
import React, { useEffect, useState } from "react";
import Button from "./ui/button";

type theme = "dark" | "light" | "system";

const ChangeTheme = () => {
	const [theme, setTheme] = useState<theme>("dark");

	useEffect(() => {
		const htmlElement = document.getElementsByTagName("html")[0]; // Get the first <html> element

		if (theme === "system") {
			// Use system preference
			const prefersDarkMode = window.matchMedia(
				"(prefers-color-scheme: dark)"
			).matches;
			setTheme(prefersDarkMode ? "dark" : "light");
		} else {
			htmlElement.classList.remove("light", "dark");
			htmlElement.classList.add(theme);
		}
	}, [theme]);

	return (
		<div className="container">
			<h3 className="text-center capitalize">dark theme and light theme</h3>
			<div className=" flex gap-2 justify-center">
				<Button onClick={() => setTheme("light")}>Light</Button>
				<Button onClick={() => setTheme("dark")}>Dark</Button>
				<Button onClick={() => setTheme("system")}>System</Button>
			</div>
		</div>
	);
};

export default ChangeTheme;
`
_____________________________________________________
